### ISSUE TITLE
Sql Management Objects async overloads

### ISSUE THREAD

**Issue by MgSam** [microsoft/sqlmanagementobjects#72]
(Please let me know if this is the wrong repo)

I was pleasantly surprised to find SMO is staying up-to-date with the latest features of SQL Server and is available on .NET Core. 

One area in which its sorely lacking, however is in async support. `async`/`await` has been in the C# language since 2012; Async overloads are crucial in nearly every framework doing I/O in .NET.

SMO should add Async overloads for any method call that results in a call to the database. This is important so that a thread is not just sitting there, blocked and idle, while waiting on an external operation.

**Comment by shueybubbles** [comment]
There are no plans to change the SMO model to be anything but synchronous. The abstraction in the object model is meant to hide database queries from the caller.

**Comment by Piedone** [comment]
I'm baffled as well. I can understand that it's not a priority right now but unless SMO is or will be deprecated, providing async APIs must eventually be done.

Perhaps if the project would be open-source, people would even contribute this?

**Comment by Pilchard123** [comment]
The reasoning given doesn't even hold water. There are methods that explicitly say they're for making queries: [Database.ExecuteWithResults](https://docs.microsoft.com/en-us/dotnet/api/microsoft.sqlserver.management.smo.database.executewithresults) and [Database.ExecuteNonQuery](https://docs.microsoft.com/en-us/dotnet/api/microsoft.sqlserver.management.smo.database.executenonquery). I don't see how you can hide the fact that a method that is for making database queries, that you pass a database query, and (in the case of ExecuteWithResults) gives you the result of making that database query... makes a database query. It's like trying to hide the fact that labelled "bananas" and that you just put some bananas into contains bananas.

**Comment by dzsquared** [comment]
Hi folks! I'm glad to see interest in SMO. 

Just a reminder that this repo has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/)

In particular, I'd like to encourage this conversation to continue while being mindful of the standards:

> Examples of behavior that contributes to a positive environment for our community include:.
> Demonstrating empathy and kindness toward other people
> Being respectful of differing opinions, viewpoints, and experiences
> Giving and gracefully accepting constructive feedback
> Accepting responsibility and apologizing to those affected by our mistakes, and learning from the experience
> Focusing on what is best not just for us as individuals, but for the overall community

**Comment by dzsquared** [comment]
@MgSam @Piedone @Pilchard123 
So I can better understand and advocate for async support within SMO, can you elaborate a bit on your specific scenarios and the utilization of SMO where it would be beneficial to have async/callback support?

**Comment by mdsharpe** [comment]
2026 and still no async methods on SMO! 😲

**Comment by shueybubbles** [comment]
I'd be happy to unleash an AI model on the problem - what I need from the community is a clearer specification than just "you need async!".  Please contribute some code snippets showing how you'd like the public interface to be revised. Are folks just asking for the highest level wrapper objects like `Scripter` and `ScriptMaker` and `RestorePlan` to offer asynchronous methods while the underlying SFC objects like `Request` and `Enumerator` remain synchronous? Do you want asynchronous collection initialization like `Database.Tables.InitializeAsync()` which would fully populate the `Table` objects in the collection while keeping the default enumeration of collections synchronous?

**Comment by mdsharpe** [comment]
I can't speak for anyone else, but I ended up here after finding that `ServerConnection.ExecuteNonQuery` has no async equivalent.

**Comment by shueybubbles** [comment]
@mdsharpe @Pilchard123 progress has been made with 181.19. If you have some feedback on how it's helpful or not I'd appreciate it.

**Comment by mdsharpe** [comment]
I noticed async methods appear recently! Was a pleasant surprise. Looked good!
